corehq/motech/repeaters/dbaccessors.py

```diff
@@ -74,34 +77,41 @@ def get_overdue_repeat_record_count(overdue_threshold=datetime.timedelta(minutes
     return results['value'] if results else 0
 
 
-def _get_startkey_endkey_all_records(domain, repeater_id=None, state=None):
+def _get_startkey_endkey_all_records(domain, repeater_id=None, state=None, datespan=None):
     kwargs = {}
 
-    if repeater_id and not state:
-        kwargs['endkey'] = [domain, repeater_id]
-        kwargs['startkey'] = [domain, repeater_id, {}]
-    elif repeater_id and state:
+    if datespan:
+        # reversed because descending:True
+        enddatekey = json_format_datetime(datespan.startdate) if datespan.startdate else ""
+        startdatekey = json_format_datetime(datespan.enddate) if datespan.enddate else ""
+        kwargs['endkey'] = [domain, repeater_id, state, enddatekey]
+        kwargs['startkey'] = [domain, repeater_id, state, startdatekey]
+    elif state:
         kwargs['endkey'] = [domain, repeater_id, state]
-        kwargs['startkey'] = [domain, repeater_id, state, {}]
-    elif not repeater_id and state:
-        kwargs['endkey'] = [domain, None, state]
-        kwargs['startkey'] = [domain, None, state, {}]
-    elif not repeater_id and not state:
-        kwargs['endkey'] = [domain, None]
-        kwargs['startkey'] = [domain, None, {}]
+        kwargs['startkey'] = [domain, repeater_id, state, None]
+    elif repeater_id:
+        kwargs['endkey'] = [domain, repeater_id]
+        kwargs['startkey'] = [domain, repeater_id, None, None]
+    else:
+        kwargs['endkey'] = [domain, None, None]
+        kwargs['startkey'] = [domain, None, None, None]
```

Comment:
what happens if you want to filter by 'repeater_id' AND by 'state'?

I think the previous code catered for that explicitly but now it's implicity based on the values passed in which are used in the keys correct?

This function could really use a docstring.

Reply:
> I think the previous code catered for that explicitly but now it's implicity based on the values passed in which are used in the keys correct?

Yes, that's right. The previous one iterated all combinations of the keys, the simplified approach I changed start from the last key to first key which allows implicit None values for the key that follows each key. Otherwise, since I added a 4th key to filter on, we would need about 4 times the previous condition blocks, which would be confusing.

> what happens if you want to filter by 'repeater_id' AND by 'state'?

Effectively the same as previous, the previous condition for that case [was](https://github.com/dimagi/commcare-hq/pull/32626/files#diff-3a8c7e32fc20d66cf08a263be8665458fc24303fb74500e26a1e2ae35cb9b1f8L83)

```
elif repeater_id and state:
        kwargs['endkey'] = [domain, repeater_id, state]
        kwargs['startkey'] = [domain, repeater_id, state, {}]
```

and the current condition is the same 

```
    elif state:
        kwargs['endkey'] = [domain, repeater_id, state]
        kwargs['startkey'] = [domain, repeater_id, state, None]
```
(I will update None to {})

Yes, I will add a docstring.